tests/unit/hooks/test_snapper.py

```diff
@@ -0,0 +1,265 @@
+                "/bar/.snapshots/1337/snapshot",
+                "/baz/.snapshots/13337/snapshot",
+                "/quaz",
+            ],
+        ),
+    ),
+)
+def test_use_snapshots_include_exclude(config, src_dirs_in, expected):
+    log_prefix = ''
+
+    flexmock(module).should_receive('execute_command_and_capture_output').with_args(
+        ["snapper", "--jsonout", "list-configs"]
+    ).and_return(
+        json.dumps(
+            {
+                "configs": [
+                    {
+                        "config": "foo",
+                        "subvolume": "/foo",
+                    },
+                    {
+                        "config": "bar",
+                        "subvolume": "/bar",
+                    },
+                    {
+                        "config": "baz",
+                        "subvolume": "/baz",
+                    },
+                ],
+            }
+        )
+    )
+    flexmock(module).should_receive('execute_command_and_capture_output').with_args(
+        ["snapper", "--jsonout", "-c", "foo", "list", "--disable-used-space"]
+    ).and_return(
+        json.dumps(
+            {
+                "foo": [
+                    {
+                        "number": 17,
+                    },
+                ]
+            }
+        )
+    )
+    flexmock(module).should_receive('execute_command_and_capture_output').with_args(
+        ["snapper", "--jsonout", "-c", "bar", "list", "--disable-used-space"]
+    ).and_return(
+        json.dumps(
+            {
+                "bar": [
+                    {
+                        "number": 1337,
+                    },
+                ]
+            }
+        )
+    )
+    flexmock(module).should_receive('execute_command_and_capture_output').with_args(
+        ["snapper", "--jsonout", "-c", "baz", "list", "--disable-used-space"]
+    ).and_return(
+        json.dumps(
+            {
+                "baz": [
+                    {
+                        "number": 13337,
+                    },
+                ]
+            }
+        )
+    )
+    flexmock(module.Path).should_receive("exists").and_return(True)
+
+    src_dirs_after = module.prepare_source_directories(config, log_prefix, src_dirs_in)
+    assert set(src_dirs_after) == set(expected)
+
+
+def test_no_config():
+    log_prefix = ''
+
+    flexmock(module).should_receive('execute_command_and_capture_output').once().with_args(
+        ["snapper", "--jsonout", "list-configs"]
+    ).and_return(
+        json.dumps(
+            {
+                "configs": [
+                    {
+                        "config": "bar",
+                        "subvolume": "/bar",
+                    },
+                ],
+            }
+        )
+    )
+
+    src_dirs_before = ["/foo"]
+    with pytest.raises(ValueError) as e_info:
+        module.prepare_source_directories({"include": ["/foo"]}, log_prefix, src_dirs_before)
+    assert "could not be found" in str(e_info)
```

Comment:
This assertion might need indenting under the `with` so `e_info` is always set? For instance, does it get set if there's no exception? I don't think there's risk of exiting the function cleanly if there's no `ValueError` exception because of that `pytest.raises()`.

Reply:
I have always written my tests like this, and I think due to how python handles variable scoping, this is functionally equivalent. However your approach seems to be more clean, so I changed it.